### PR TITLE
Generate: assistant should be greedy in assisted decoding

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Restores a previous behavior -- forces the assistant in assisted decoding to do greedy decoding.

While providing [this answer](https://github.com/huggingface/transformers/issues/30608#issuecomment-2107585243), I noticed that we dropped a property in the refactor that was present in the original implementation: the assistant should be greedy.

Why? We want the assistant to maximize the number of matches with the original model, to result in the maximum speedup. Remember: the main model always calls the shots, the output text is never changed regardless of the flags we set for the assistant. By forcing the assistant to be greedy, we pick its most likely tokens. If the assistant model has a distribution aligned with the main model, these tokens will also be the main model's most likely tokens, increasing the number of matches.

When running my original benchmarks (https://github.com/gante/huggingface-demos/tree/main/experiments/faster_generation), this resulted in a ~20% speedup @ temperature=0.5 over the current version of the code.

### CODE REVIEW

File: src/transformers/generation/configuration_utils.py [hunk 1]
```diff
@@ -496,6 +496,11 @@ def validate(self, is_init=False):
                     greedy_wrong_parameter_msg.format(flag_name="top_p", flag_value=self.top_p),
                     UserWarning,
                 )
+            if self.min_p is not None:
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:499]: This check was missing 👼

File: src/transformers/generation/candidate_generator.py [hunk 2]
```diff
@@ -150,6 +150,12 @@ def __init__(
         self.generation_config.return_dict_in_generate = True
         self.generation_config.output_scores = True
 
+        # Disable sampling -- this implementation of assited generation/speculative decoding uses the assistant
```
Comment (amyeroberts) [inline src/transformers/generation/candidate_generator.py:153]:
```suggestion
        # Disable sampling -- this implementation of assisted generation/speculative decoding uses the assistant
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! 

Will this result in different outputs for the users? If so, we might want to make sure to run any slow tests. 

Question for the future: would we ever want to enable a samplng strategy for the assistant?

**Comment by gante** [comment]
@amyeroberts 

> Will this result in different outputs for the users? 

Nope :) The only difference is execution time. (The output is entirely defined by the main model and its generation config, both untouched in this PR)

> Question for the future: would we ever want to enable a sampling strategy for the assistant?

Perhaps. A potential strategy for improvement is to sample several continuations and try them all as candidates. But that would require significant changes over the current code 👀

**Comment by gante** [comment]
> > Will this result in different outputs for the users?

> Nope 

@amyeroberts my answer did not age well (#32867 -- some randomness was lost as a result of this change)

**Pull request merged**